ecl/eclcc/NEWACTIVITY.rst

```diff
@@ -0,0 +1,537 @@
+  - hqlfold
+ 
+    This is the constant folder.  At the moment the only change is to ensure that fields that are assigned constants
+    within the transform are processed correctly.  Future work could add code to optimize quantile applied to an
+    empty dataset, or selecting 1 division.
+   
+  - hqlmeta
+ 
+    Similar to the functions in hqlattr that calculate derived attributes, these functions are used to calculate how
+    the rows coming out of an activity are sorted, grouped and distributed.  It is vital to only preserve information
+    that is guaranteed to be true � otherwise invalid optimizations might be performed on the rest of the expression
+    tree.
+   
+  - reservedwords.cpp
+ 
+    A new entry indicating which category the keyword belongs to.
+
+Finally we have the changes to the parser to recognise the new syntax:
+
+* hqllex.l
+
+  This file contains the lexer that breaks the ecl file into tokens.  There are two new tokens � QUANTILE and SCORE.
+ 
+* Hqlgram.y
+
+  This file contains the grammar that matches the language.  There are two productions � one that matches the version of
+  QUANTILE with a transform and one without.  (Two productions are used instead of an optional transform to avoid
+  shift/reduce errors.)
+ 
+* hqlgram2.cpp
+
+  This contains the bulk of the code that is executed by the productions in the grammar.  Changes here include new
+  entries added to a case statement to get the text for the new tokens, and a new entry in the simplify() call.  This
+  helps reduce the number of valid tokens that could follow when reporting a syntax error.
+
+Looking back over those changes, one reflection is that there are lots of different places that need to be changed.  How 
+does a programmer know which functions need to change, and what happens if some are missed?  In this example, the 
+locations were found by searching for an activity with a similar syntax e.g., no_soapcall_ds or no_normalize.
+
+It is too easy to miss something, especially for somebody new to the code - although if you do then you will trigger a 
+runtime internal error.  It would be much better if the code was refactored so that the bulk of the changes were in one 
+place.  (See JIRA https://track.hpccsystems.com/browse/HPCC-13434 that has been added to track improving the situation.)
+
+With these changes implemented the examples from the previous pull request now syntax check.  The
+next stage in the process involves thinking through how the activity will be implemented.
+
+Quantile 4 � The engine interface.
+==================================
+The next stage in adding a new activity to the system is to define the interface between the generated code and the 
+engines.  The important file for this stage is rtl/include/eclhelper.hpp that contains the interfaces between the 
+engines and the generated code.  These interfaces define the information required by the engines to customize each
+different activity.  The changes are in commit 
```

Comment:
The required changes for this section are contained in commit (or something more than just "The changes are in" . What changes, i.e. relating to what? This will be blog post that may be read without having read the others.